File: src/main/java/com/github/joelws/release/tracker/service/release/operation/CreateReleaseServiceOperation.java
package com.github.joelws.release.tracker.service.release.operation;

import com.github.joelws.release.tracker.conversion.ReleaseDtoToReleaseAdapter;
import com.github.joelws.release.tracker.conversion.ReleaseToReleaseDtoAdapter;
import com.github.joelws.release.tracker.dto.release.ReleaseDto;
import com.github.joelws.release.tracker.entity.release.Release;
import com.github.joelws.release.tracker.handler.JsonResponse;
import com.github.joelws.release.tracker.service.ServiceHelper;
import com.github.joelws.release.tracker.service.ServiceOperation;
import com.github.joelws.release.tracker.service.release.execution.CreateReleaseServiceExecution;

import javax.ws.rs.core.Response;

public class CreateReleaseServiceOperation extends ServiceOperation<String> {
    private final ServiceHelper helper;

    private final CreateReleaseServiceExecution createReleaseServiceExecution;

    public CreateReleaseServiceOperation(ServiceHelper helper,
                                         CreateReleaseServiceExecution createReleaseServiceExecution) {
        this.helper = helper;
        this.createReleaseServiceExecution = createReleaseServiceExecution;
    }

    @Override
    protected Response delegate(String json) {

        final ReleaseDtoToReleaseAdapter adapter = helper.getFactory().getImpl(ReleaseDtoToReleaseAdapter.class);

        final Release result = createReleaseServiceExecution
                .execute(adapter.adapt(helper.getJsonAdapter().getObjectFromJson(json, ReleaseDto.class)));

        Response response;

        if (result != null) {
            final ReleaseToReleaseDtoAdapter adapterForResponse = helper.getFactory().getImpl(ReleaseToReleaseDtoAdapter.class);
            final ReleaseDto adaptedResult = adapterForResponse.adapt(result);
            response = helper.getRestResponseBuilder().build(200, adaptedResult);
        } else {
            response = helper.getRestResponseBuilder().build(new JsonResponse(442, "Release already exists"));
        }
        return response;
    }
}

File: src/main/java/com/github/joelws/release/tracker/service/release/operation/CreateReleaseServiceOperation.kt
package com.github.joelws.release.tracker.service.release.operation;

import com.github.joelws.release.tracker.conversion.ReleaseDtoToReleaseAdapter
import com.github.joelws.release.tracker.conversion.ReleaseToReleaseDtoAdapter
import com.github.joelws.release.tracker.dto.release.ReleaseDto
import com.github.joelws.release.tracker.response.RestResponse.BadRequest
import com.github.joelws.release.tracker.response.RestResponse.SuccessWithEntity
import com.github.joelws.release.tracker.response.build
import com.github.joelws.release.tracker.service.ServiceHelper
import com.github.joelws.release.tracker.service.ServiceOperation
import com.github.joelws.release.tracker.service.release.execution.CreateReleaseServiceExecution
import javax.ws.rs.core.Response

open class CreateReleaseServiceOperation(private val helper: ServiceHelper,
                                         private val createReleaseServiceExecution: CreateReleaseServiceExecution) : ServiceOperation<String> {

    override fun delegate(param: String?): Response {

        val adapter = helper.factory.getImpl(ReleaseDtoToReleaseAdapter::class.java)

        val result = createReleaseServiceExecution
                .execute(adapter.adapt(helper.jsonAdapter.getObjectFromJson(param, ReleaseDto::class.java)))

        return if (result != null) {
            val adapterForResponse = helper.factory.getImpl(ReleaseToReleaseDtoAdapter::class.java)
            val adaptedResult = adapterForResponse.adapt(result)
            SuccessWithEntity(adaptedResult).build()
        } else {
            BadRequest("Release already exists").build()
        }
    }
}
